lisa/_assets/kmodules/lisa/fs.c

```diff
@@ -0,0 +1,755 @@
+ *                    umode_t);
+ */
+
+static int lisa_fs_syscall_mkdir(struct mnt_idmap *idmap,
+				 struct inode *inode, struct dentry *dentry,
+				 umode_t mode)
+{
+	struct dentry *my_dentry;
+	struct lisa_cfg *cfg;
+	int ret;
+
+	cfg = allocate_lisa_cfg(dentry->d_name.name);
+	if (!cfg)
+		return -ENOMEM;
+
+	mutex_lock(&interface_lock);
+
+	my_dentry = lisa_fs_create_single(dentry->d_parent, dentry->d_name.name,
+					  &simple_dir_inode_operations,
+					  &simple_dir_operations,
+					  S_IFDIR | mode, cfg);
+	if (!my_dentry)
+		goto error;
+
+	ret = init_lisa_cfg(cfg, &cfg_list, my_dentry);
+	if (ret)
+		return -ENOMEM;
+
+	lisa_fs_create_files(my_dentry, false, cfg);
+	mutex_unlock(&interface_lock);
+	return 0;
+
+error:
+	free_lisa_cfg(cfg);
+	mutex_unlock(&interface_lock);
+	return ret;
+}
+
+void lisa_fs_remove(struct dentry *dentry)
+{
+	simple_recursive_removal(dentry, NULL);
+}
+
+static int lisa_fs_syscall_rmdir(struct inode *inode, struct dentry *dentry)
+{
+	struct lisa_cfg *cfg;
+	cfg = inode->i_private;
+
+	inode_unlock(inode);
+	inode_unlock(d_inode(dentry));
+
+	mutex_lock(&interface_lock);
+
+	cfg = find_lisa_cfg(dentry->d_name.name);
+	if (!cfg)
+		pr_err("Failed to find config: %s\n", dentry->d_name.name);
+
+	free_lisa_cfg(cfg);
+
+	mutex_unlock(&interface_lock);
+
+	inode_lock_nested(inode, I_MUTEX_PARENT);
+	inode_lock(d_inode(dentry));
+
+	return 0;
+}
+
+const struct inode_operations lisa_fs_dir_inode_operations = {
+	.lookup = simple_lookup,
+	.mkdir = lisa_fs_syscall_mkdir,
+	.rmdir = lisa_fs_syscall_rmdir,
+};
+
+/////////////////////////////////////
+//           Main files
+/////////////////////////////////////
+
+static struct dentry *
+lisa_fs_create_single(struct dentry *parent, const char *name,
+		      const struct inode_operations *i_ops,
+		      const struct file_operations *f_ops, umode_t mode,
+		      void *data)
+{
+	struct dentry *dentry;
+	struct inode *inode;
+
+	dentry = d_alloc_name(parent, name);
+	if (!dentry)
+		return NULL;
+	inode = lisa_fs_create_inode(parent->d_sb, mode);
+	if (!inode) {
+		dput(dentry);
+		return NULL;
+	}
+
+	if (mode & S_IFREG) {
+		inode->i_fop = f_ops;
+	} else {
+		inode->i_op = i_ops;
+		inode->i_fop = f_ops;
+	}
+	inode->i_private = data;
+	d_add(dentry, inode);
+	if (mode & S_IFDIR) {
+		inc_nlink(d_inode(parent));
+		inc_nlink(inode);
+	}
+
+	return dentry;
+}
+
+/* Called with interface_lock */
+static int
+lisa_fs_create_files(struct dentry *parent, bool is_top_level, struct lisa_cfg *cfg)
+{
+	struct feature_param_entry *entry;
+	struct feature *feature;
+
+	entry = allocate_feature_param_entry();
+	if (!entry)
+		return -ENOMEM;
+
+	init_feature_param_entry(entry, cfg, &lisa_features_param);
+
+	/* set_features: enable a feature - RW. */
+	if (!lisa_fs_create_single(parent, "set_features",
+			      NULL,
+			      &lisa_param_feature_fops,
+			      S_IFREG | S_IRUGO | S_IWUGO, entry)) {
+		free_feature_param_entry(entry);
```

Comment:
It escaped me why this is the only case when the earlier allocated and initialized entry gets properly cleaned upon failure. Why this is not being done for other cases ?  Will that be covered by free_lisa_cfg somehow ? Either way, unless I missed smth it would be best to either clean thing up on each failure, or document why it is not needed.

Reply:
The call to `free_feature_param_entry()` should not be necessary actually. If something goes wrong, the caller should call `free_lisa_cfg()`, which will go over the list of entries and free them.
I'll add a comment in that sense.

Reply:
OK, got it - thanks!